polynomial.hpp

```diff
@@ -0,0 +1,220 @@
+        auto operator*=(const polynomial &p) -> polynomial &{
+            coefficients.resize(p.degree + degree + 1, T{0});
+            for (std::size_t i = 0; i < degree; ++i)
+            {
+                for (std::size_t j = 0; j < p.degree; ++j)
+                {
+                    coefficients[i+j] += coefficients[i] * p.coefficients[j];
+                }
+            }
+            degree += p.degree;
+            return *this;            
+        }
+        auto operator/=(const polynomial &p) -> polynomial & {
+            if (p.degree == 0 && p.coefficients[0] == T{0}) {
+                throw std::invalid_argument("Division by zero polynomial");
+            }else if(p.degree > 1){
+                throw std::invalid_argument("Division by invalid polynomial");
+            }
+            coefficients.resize(degree, T{0});
+            degree -= 1;
+            for(std::size_t i; i <= degree; i++){
+                coefficients[i] /= p.coefficients[0];
+                if( i+1 <= degree){
+                    coefficients[i+1] -= coefficients[i] * p.coefficients[1];
+                }
+            }
+            return *this;
+        }
+        auto operator%=(const polynomial &p) -> polynomial & {
+            if (p.degree == 0 && p.coefficients[0] == T{0}) {
+                throw std::invalid_argument("Division by zero polynomial");
+            }
+            polynomial quotient;
+            polynomial remainder = *this;
+            quotient = *this / p;
+            remainder -= quotient * p;
+            *this = remainder;
+            return *this;
+        }
+      public:
+        auto operator()(T value) {
+        	T result = 0;
+            for (int i = this->degree; i >= 0; --i) {
+                result = result * value + this->coefficients[i];
+            }
+            return result;
+        }
+    };
+    template <typename T>
+    inline auto operator+(polynomial<T> p, polynomial<T> q) {
+        p += q;
+        return p;
+    }
+    template <typename T>
+    inline auto operator-(polynomial<T> p, polynomial<T> q) {
+    	p -= q;
+    	return p;
+    }
+    template <typename T>
+    inline auto operator*(polynomial<T> p, polynomial<T> q) {
+    	p *= q;
+    	return p;
+    }
+    template <typename T>
+    inline auto operator/(polynomial<T> p, polynomial<T> q) {
+    	p /= q;
+    	return p;
+    }
+    template <typename T>
+    inline auto operator%(polynomial<T> p, polynomial<T> q) {
+    	p %= q;
+    	return p;
+    }
+    template <typename T>
+    inline auto divide(polynomial<T> p, polynomial<T> q)
+    ->std::pair<polynomial<T>, polynomial<T>> {
+    	polynomial<T> quotient, remainder;
+        quotient = p / q;
+        remainder = p % q;
+        return std::make_pair(quotient, remainder);
+    }
+    template <typename T>
+    inline auto root_rational_candidates(polynomial<T> p) -> std::vector<T>{
+        size_t P = p.coefficients[p.degree];
+        size_t Q = p.coefficients[0];
```

Comment:
use descriptive names e.g. P is called the leading coefficient, you can use for example `constant_term` for `Q`